tests/gtest/cppapi_test.cpp

```diff
@@ -0,0 +1,104 @@
+#include <map>
+
+#include "ctypes.h"
+#include "cppapi.hpp"
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+namespace
+{
+    const std::string SAMPLES_10 = "file://10_samples_default.vds";
+    const std::string CREDENTIALS = "";
+
+    std::map<float, std::vector<float>> samples_10_data(float fill)
+    {
+        std::map<float, std::vector<float>> points;
+        points[110] = {-4.5, -3.5, -2.5, -1.5, -0.5, 0.5, 1.5, 2.5, 3.5, 4.5};
+        points[111] = {4.5, 3.5, 2.5, 1.5, 0.5, -0.5, -1.5, -2.5, -3.5, -4.5};
+        points[310] = {25.5, -14.5, -12.5, -10.5, -8.5, -6.5, -4.5, -2.5, -0.5, 25.5};
+        points[311] = {25.5, 0.5, 2.5, 4.5, 6.5, 8.5, 10.5, 12.5, 14.5, 25.5};
+        points[510] = {25.5, 4.5, 8.5, 12.5, 16.5, 20.5, 24.5, 20.5, 16.5, 8.5};
+        points[511] = {25.5, -4.5, -8.5, -12.5, -16.5, -20.5, -24.5, -20.5, -16.5, -8.5};
+        points[fill] = {fill};
+        return points;
+    }
+
+    class HorizonTest : public ::testing::Test
+    {
+    protected:
+        void SetUp() override
+        {
+            handle = make_datahandle(SAMPLES_10.c_str(), CREDENTIALS.c_str());
```

Comment:
I am not following. Do you suggest I implement a new constructor specially for this? Because make_datahandle does a bit more than just new Datahandle.

Reply:
This is what I suggested: No `new`, hence no `delete`, hence no Teardown. But since the constructer of DataHandle is private (which I'm not sure I agree it should be now that we have tests), we can't to this without changing DataHandle. I'll leave any decisions on this to you

```
void SetUp() override {
    OpenVDS::Error error;
    auto handle = OpenVDS::Open(SAMPLES_10.c_str(), CREDENTIALS.c_str(), error);
    if(error.code != 0) {
        throw std::runtime_error("Could not open VDS: " + error.string);
    }
    handle = DataHandle(std::move(handle));
}

Reply:
I think I have to pass.
My main wish here is to keep Open::open - error - etc local to the datahandle. I do not want to repeat it again and again and again. Which means that I still want a friend method that would be able to construct the object.
And I understand that for this we need "new" anyway.

Additionally something in that code doesn't compile even after changes (it complains it lacks copy-constructor).

(And it also unclear how it can know about Openvds::Error  and so. We removed openvds from include directories, but it still works even though tests have no explicit knowledge about openvds?? I don't understand how cmake works :smile: )

Reply:
> (And it also unclear how it can know about Openvds::Error and so. We removed openvds from include directories, but it still works even though tests have no explicit knowledge about openvds?? I don't understand how cmake works 😄 )

Good question! It's because the tests link to the `vdscore` target which again _publically_ link to openvds. _public_ here mean that if the dependency to openvds in `vdscore` is visible to anyone linking towards it (`vdscoretest`). See https://github.com/achaikou/vds-slice/blob/0896d1af939dfb51e486832671ade5dd2b76baa9/internal/core/CMakeLists.txt#L23

So what is actually more correct in this case, is to change that line to private. Then the tests doesn't find openvds through vdscore and instead has to add openvds to it's own `target_link_libraries`

How the three targets currently depend on each other
```
vdscoretest -> vdscore -> openvds
```

How they depend on each other in my suggestion
```
vdscore -> openvds

vdscoretest -> vdscore
                     -> openvds
```